### DISCUSSION TITLE
Is there an easy way to make Pico.css NAV responsive?

### DISCUSSION THREAD

**Discussion by bjazmoore** [picocss/pico#343] [Help]
I started playing around with making the NAV work responsively.  I was able to hide the menu items and slap on a hamburger menu.  Obviously I would need some Javascript to make the UL/LI menu items display as a stacked menu item.  I thought about slapping an ASIDE tag into the NAV, but that does not seem to do what I want.  My extra CSS and custom classes started getting really complex and I thought "this is not the Pico.css way".  

Anyone have some examples?  I can post my code if you want it.

Thanks

**Comment by walterdavis** [comment]
Why would you need JavaScript to change the list from horizontal (li set to display: inline) to vertical (li set to display: block)? That can live in your viewport-specific part of the CSS, along with the visible hamburger etc.

Walter

**Reply by bjazmoore** [reply]
In the past I have used javascript to inject a class that makes the menu items visible when the hamburger menu is clicked and the hides the by removing the class when clicked again.  I am not aware of how to do that with just CSS.  Regardless - what I see when attempting this is a stacked menu but the top item appears on the navbar and not right below it.  

Thanks

**Comment by wwmoraes** [comment]
@bjazmoore you need neither JS nor an `aside` element for that:

```css
nav, nav ul {
  flex-direction: column;
}

@media (min-width: 576px) {
  nav, nav ul {
    flex-direction: row;
  }
}
```

[Here's a sample of it in action.](https://jsbin.com/wamovex/edit?html,css,output) This will make both the `nav` and its child `ul` elements to change the flow direction to column on narrow viewports 😉 

Note that I used the latest version as of date (v2.0.6). It works for any version that sets `display: flex` to `nav, nav ul` .

As for the burger menu, you can create one with its toggling effect in plain CSS using a styled label + checkbox input + rules that depend on `input[type="checkbox"]:checked`. I applied [this trick on my own website](https://github.com/wwmoraes/site/blob/7dcd9c7c7bb5925295ec0be06b4ee5b4a90685d9/themes/m10d/assets/css/components/_toggle.scss#L83-L85) some time ago using the subsequent sibling selector (`~`) to target the menu. This trick requires the input to exist outside of the target element, as its previous sibling. A more modern approach (2023+) may have the checkbox within the menu container + `:has()` pseudo-class hiding menu entries/sections instead.

[Here's another sample](https://jsbin.com/wasarok/edit?html,css,output) that contains the responsive nav, burger menu to toggle menu items on narrow viewports using the latter approach and some animations to make it smooth.

P.S.: Sorry to necropost, I only started using pico.css a week ago and I ended up looking for a potential pre-built way to enable this, leading me to this question and solution. Cheers!

**Reply by Yohn** [reply]
@wwmoraes Nice, I like how you were able to set that up! I found pico a little over a month ago or so, and started adding a bunch of fixes, pull requests from here and new components to [my fork](https://github.com/Yohn/PicoCSS/) that also had a hamburger menu but yours is definitely better. It would be an awesome improvement, if you'd like to create a pull request?

**Reply by wwmoraes** [reply]
@Yohn Glad to see my hacks aren't that hack-ish after all 😄 

It'd be my pleasure to contribute. A bit more on that off-topic: I've also extended pico.css with some niceties while playing around, such as support for `<meter>`. I'll come back to this once I finish my original plan to give my site a new theme made from scratch using pico.

**Reply by Yohn** [reply]
@wwmoraes can't wait to see what else you've added! Be sure to check out my fork cause some things might already be fixed, it added there! It looks like I'll be working on another site now as well, where I will be using PicoCSS, so I'm sure I'll be adding some more tweaks within the coming weeks too!

**Reply by wwmoraes** [reply]
> can't wait to see what else you've added!

@Yohn wait no more then: I've released my new site theme at last. You can find my pico tweaks [here](https://github.com/wwmoraes/site/tree/master/themes/pico/assets/css/pico) 😃 

You may also find some other interesting styles outside of that folder. I separated pure pico adjustments from my specific needs/preferences.

> Be sure to check out my fork cause some things might already be fixed, it added there!

Thank you, I just did. I see you added support for shadom DOM styling, a grid system, form novalidate support, burger menu, tabs and timeline, aside from adjustments in a bunch of existing styles. Plenty of good work I must say!

For now I will stay on the vanilla pico though. I don't need a grid system, forms or even the burger menu anymore that would justify the ~40% increase in minified CSS footprint. In fact I opted for pico precisely for its motto present in their homepage: `Write HTML, Add Pico CSS, and Voilà!`. I appreciate its semantic-first approach and how clean and uncluttered my markup is now, making it much more maintainable.

Speaking of which, here's an honest question: do you believe adding a grid system for instance is within pico's original goals? Perhaps you could offer those "classful" extras as external extensions for those that need something less _pico_ and more _nano_/_micro_; or if you're into classic references, _pico++_ 😆 

P.S.: Yes I know I can disable modules using SCSS. I need to explore hugo's modules, git submodules or another solution to use the upstream sources next instead of the shamelessly copy-pasted upstream CSS file I use right now 😅

**Reply by Yohn** [reply]
Hey @wwmoraes I actually answered some of your concerns in [this discussion](https://github.com/Yohn/PicoCSS/discussions/33) about keeping Pico smaller.

Honestly, I do not think Pico needs a full row / column grid system at all, and I would like to try and incorporate the CSS `grid-template` to set the basic side panels in a template easily. 

Most of the enhancements I've added to my fork I focused on not having or using additional classes for those elements, but yes, the rows, and timeline are class intensive, and those were, of course, the first things I added because I needed them for a client  at that time. I put notes in the [_settings.scss](https://github.com/Yohn/PicoCSS/blob/main/scss/_settings.scss#L123) file for some plans I have for V3 last month which include removing the rows, and disabling the timeline by default or possibly removing it completely and making a new "nano" project for such features.

Off topic kind of, I'm working on a project now where I'm using PicoCSS and the client its for is Nano Farms, so thats funny you mentioned nano in your comment too haha. It must be a sign!

And onto what you did! I love that you brought the `<menu>` tag into your project!! I've been exploring ways to do that as well What are you using that for thats different than the current `<nav>`? I tried browsing your site for where you did use the menu tag but was unable to find it. I'd be interested in seeing how you did that, and the meter that you discussed!

**Reply by wwmoraes** [reply]
> What are you using that for thats different than the current `<nav>`? I tried browsing your site for where you did use the menu tag but was unable to find it. I'd be interested in seeing how you did that, and the meter that you discussed!

My bad, I didn't use them in the current production pages. I have a [style guide in my staging environment](https://staging.artero.dev/internal/style/). Scroll to the bottom and you'll see both meter and menu sections with samples 😉 

In short, `meter` got a visual revamp to replace user-agent glossy/corny styles with pico's and to add responsiveness.

Menu I shamelessly re-used the current pico styles for `nav[role="menu"]`, then added [responsiveness using flex-direction](https://github.com/wwmoraes/site/blob/master/themes/pico/assets/css/pico/_nav.scss#L24-L45). Lower resolutions grow column-wise; higher resolutions grow row-wise. 

Semantically speaking a `menu`/`nav[role="menu"]` alone is no different from any other nav. The real difference is when you want to toggle it. To enable toggling a menu needs a [label+checkbox child](https://github.com/wwmoraes/site/blob/staging/themes/pico/assets/css/_toggle.scss#L73). Then you get:
- toggles `role="menuitem"` elements
- uses `role="menubar"` as "always visible" elements (no styles needed so far)
- reverts to a standard, non-toggling menu at higher resolutions

There's one caveat: it uses `max-height` with `overflow-y: visible` as an alternative to absolute positioning, so it does not adjust automatically. It sounds acceptable to me; YMMV.

**Reply by Yohn** [reply]
Oh man, and you got a style guide! With your permission, I'd like to copy/paste that and bring it over to my fork I started working on a cheat sheet but didn't get as far as you have with your style guide and got distracted with one of the many projects going on.

Both the meter and menu components you added are awesome! I'd love to get those added to my fork as well. The menu component you got I feel like is way better than what I got going on. 

Have you tried added a dropdown component to the menu items?

The green and red colors in the meter, is that the same color as the `<ins>` and `<del>` / valid and invalid form elements?

Loving what you did, great work!

**Reply by Yohn** [reply]
I just added a dropdown menu to the `<menu>` hamburger style element you got set up, and it looks like it worked just fine!

**Reply by wwmoraes** [reply]
> Oh man, and you got a style guide! With your permission, I'd like to copy/paste that and bring it over to my fork I started working on a cheat sheet but didn't get as far as you have with your style guide and got distracted with one of the many projects going on.

Feel free to copy and use it: its MIT-licensed for that purpose 😄 

> The green and red colors in the meter, is that the same color as the `<ins>` and `<del>` / valid and invalid form elements?

[The latter](https://github.com/wwmoraes/site/blob/66247579e8cbc0ac03cf3e9b1cbacce471fe5979/themes/pico/assets/css/pico/_meter.scss#L2-L5). Sub-optimal I picked the first yellow tone that fits (subjectively) in that color palette.

> Have you tried added a dropdown component to the menu items?
>
> I just added a dropdown menu to the `<menu>` hamburger style element you got set up, and it looks like it worked just fine!

Thanks for testing! I designed it without floats or absolute positioning precisely with the goal that it'd "just work" and "not break anything else". I'm glad to see it lives up to my expectations 😎

**Reply by mnordhaus** [reply]
> Hey @wwmoraes I actually answered some of your concerns in [this discussion](https://github.com/Yohn/PicoCSS/discussions/33) about keeping Pico smaller.
> 
> Honestly, I do not think Pico needs a full row / column grid system at all, and I would like to try and incorporate the CSS `grid-template` to set the basic side panels in a template easily.
> 
> Most of the enhancements I've added to my fork I focused on not having or using additional classes for those elements, but yes, the rows, and timeline are class intensive, and those were, of course, the first things I added because I needed them for a client at that time. I put notes in the [_settings.scss](https://github.com/Yohn/PicoCSS/blob/main/scss/_settings.scss#L123) file for some plans I have for V3 last month which include removing the rows, and disabling the timeline by default or possibly removing it completely and making a new "nano" project for such features.
> 
> Off topic kind of, I'm working on a project now where I'm using PicoCSS and the client its for is Nano Farms, so thats funny you mentioned nano in your comment too haha. It must be a sign!
> 
> And onto what you did! I love that you brought the `<menu>` tag into your project!! I've been exploring ways to do that as well What are you using that for thats different than the current `<nav>`? I tried browsing your site for where you did use the menu tag but was unable to find it. I'd be interested in seeing how you did that, and the meter that you discussed!



> Hey @wwmoraes I actually answered some of your concerns in [this discussion](https://github.com/Yohn/PicoCSS/discussions/33) about keeping Pico smaller.
> 
> Honestly, I do not think Pico needs a full row / column grid system at all, and I would like to try and incorporate the CSS `grid-template` to set the basic side panels in a template easily.
> 
> Most of the enhancements I've added to my fork I focused on not having or using additional classes for those elements, but yes, the rows, and timeline are class intensive, and those were, of course, the first things I added because I needed them for a client at that time. I put notes in the [_settings.scss](https://github.com/Yohn/PicoCSS/blob/main/scss/_settings.scss#L123) file for some plans I have for V3 last month which include removing the rows, and disabling the timeline by default or possibly removing it completely and making a new "nano" project for such features.
> 
> Off topic kind of, I'm working on a project now where I'm using PicoCSS and the client its for is Nano Farms, so thats funny you mentioned nano in your comment too haha. It must be a sign!
> 
> And onto what you did! I love that you brought the `<menu>` tag into your project!! I've been exploring ways to do that as well What are you using that for thats different than the current `<nav>`? I tried browsing your site for where you did use the menu tag but was unable to find it. I'd be interested in seeing how you did that, and the meter that you discussed!

@Yohn I am about to register nanocss.org > PicoCSS in its core but with extensions and recipes for larger projects ... But I doubt to have enough capacity at the moment ;-).

**Reply by arderyp** [reply]
What am I missing?  I am pulling 

`<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2/css/pico.min.css">`

and using the exact style guide example:

```
<menu>
      <label data-role="burger"><input type="checkbox"></label>
      <ul role="menubar">
        <li><strong>Acme Corp</strong></li>
      </ul>
      <ul role="menuitem">
        <li><a href="#">About</a></li>
        <li><a href="#">Services</a></li>
        <li><a href="#">Products</a></li>
      </ul>
    </menu>
```

and its not collapsing?

**Reply by suchislife801** [reply]
## Pico.css Responsive Nav Example

Here's an example in Pure JavaScript and very minimal CSS while staying close to the source.
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Pico.css Responsive Nav</title>
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2/css/pico.min.css">
  <style type="text/css">

    /* Nav border bottom */
    nav {
      border-bottom: var(--pico-border-width) solid var(--pico-muted-border-color);
    }

    /* Hamburger menu styles */
    .nav-toggle {
      display: none;
      background: none;
      border: none;
      font-size: 1.5rem;
      cursor: pointer;
      padding: 0;
    }

    .nav-toggle:focus {
      outline: none;
      box-shadow: none !important;
    }

    /* Mobile styles */
    @media (max-width: 768px) {
      .nav-toggle {
        display: block;
      }

      nav ul:last-child {
        display: none;
        flex-direction: column;
        width: 100%;
        position: absolute;
        top: 100%;
        left: 0;
        background: var(--pico-background-color);
        border-top: var(--pico-border-width) solid var(--pico-muted-border-color);
        margin: 0;
        padding: 0;
      }

      nav ul:last-child.show {
        display: flex;
      }

      nav ul:last-child li {
        width: 100%;
        text-align: left;
      }

      nav {
        position: relative;
        flex-wrap: wrap;
      }

      nav ul:first-child {
        flex: 1;
      }
    }
  </style>
</head>
<body>
  <nav class="container">
    <ul>
      <li><strong>Brand</strong></li>
    </ul>
    <button class="nav-toggle" id="navToggle" aria-label="Toggle navigation">
      <!-- Material Design Icon : menu -->
      <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960" width="24px" fill="currentColor">
        <path d="M120-240v-80h720v80H120Zm0-200v-80h720v80H120Zm0-200v-80h720v80H120Z"/>
      </svg>
    </button>
    <ul id="navMenu">
      <li><a href="#">About</a></li>
      <li><a href="#">Services</a></li>
      <li><a href="#">Products</a></li>
      <li><a href="#">Contact</a></li>
    </ul>
  </nav>

  <main class="container">
    <h1>Responsive Navigation</h1>
    <p>Resize your browser window to see the hamburger menu appear on mobile screens.</p>
  </main>

  <script type="text/javascript">
    const navToggle = document.getElementById('navToggle');
    const navMenu = document.getElementById('navMenu');

    navToggle.addEventListener('click', () => {
      navMenu.classList.toggle('show');
    });

    // Close menu when clicking outside
    document.addEventListener('click', (e) => {
      if (!navToggle.contains(e.target) && !navMenu.contains(e.target)) {
        navMenu.classList.remove('show');
      }
    });

    // Close menu when clicking a link
    navMenu.querySelectorAll('a').forEach(link => {
      link.addEventListener('click', () => {
        navMenu.classList.remove('show');
      });
    });
  </script>
</body>
</html>
```

## Screenshots:

<img width="281" height="609" alt="Expand" src="https://github.com/user-attachments/assets/efea0eb1-bba3-48d4-bfad-824892c0f96c" />

<img width="281" height="609" alt="Collapse" src="https://github.com/user-attachments/assets/04aa8fb3-1282-4350-a6ce-32d2f3ddcd3e" />

**Comment by dday9** [comment]
You cannot make it responsive without code duplication without using some sort of JavaScript.

If you want a JavaScript-less approach, you can leverage the dialog component with a checkbox hack with minimal CSS. Just keep in mind that this would require you to duplicate the content of the menu:
```lang-html
<main>
  <nav>
    <ul>
      <li><strong>Acme Corp</strong></li>
      <li role="separator">
        <label for="checkbox-navigation">&#9776;</label>
      </li>
    </ul>
    <ul>
      <li><a href="#">About</a></li>
      <li><a href="#">Services</a></li>
      <li><a href="#">Products</a></li>
    </ul>
  </nav>
</main>

<input type="checkbox" id="checkbox-navigation" hidden>
<dialog>
  <article>
    <header>
      <nav>
        <ul>
          <li>Acme Corp</li>
        </ul>
        <ul>
          <li><label for="checkbox-navigation">&times;</label></li>
        </ul>
      </nav>
    </header>
    <aside>
      <nav>
        <ul>
          <li><a href="#">About</a></li>
          <li><a href="#">Services</a></li>
          <li><a href="#">Products</a></li>
        </ul>
      </nav>
    </aside>
  </article>
</dialog>
```
```lang-css
main > nav > ul > li[role=separator] {
  display: block;
}

main > nav > ul:not(:first-of-type) {
  display: none;
}

@media (min-width: 768px) {
  main > nav > ul > li[role=separator] {
    display: none;
  }
  main > nav > ul:not(:first-of-type) {
    display: block;
  }
}

#checkbox-navigation + dialog {
  display: none;
}

#checkbox-navigation:checked + dialog {
  display: block;
}
```

Demo: https://jsfiddle.net/2ya1k9st/

**Reply by dday9** [reply]
After giving this a bit more thought, you don't even need the checkbox hack since `command` and `commandfor` are supported:
```html
<header>
  <nav role="menubar" aria-label="Website navigation bar">
    <ul aria-label="Brand name">
      <li role="none">
        <strong>Acme Corp</strong>
      </li>
      <li role="separator">
        <button command="show-modal" commandfor="dialog-navigation" aria-label="Open navigation">
          <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960" width="24px" fill="currentColor" aria-hidden="true" focusable="false">
            <path d="M120-240v-80h720v80H120Zm0-200v-80h720v80H120Zm0-200v-80h720v80H120Z"/>
          </svg>
        </button>
      </li>
    </ul>
    <ul aria-label="Navigation">
      <li role="none"><a role="menuitem" href="#">About</a></li>
      <li role="none"><a role="menuitem" href="#">Services</a></li>
      <li role="none"><a role="menuitem" href="#">Products</a></li>
    </ul>
  </nav>
</header>

<dialog id="dialog-navigation">
  <article>
    <header>
      <button aria-label="Close" rel="prev" command="request-close" commandfor="dialog-navigation"></button>
      <p>
        <strong>Acme Corp</strong>
      </p>
    </header>
    <aside>
      <nav>
        <ul>
          <li><a href="#">About</a></li>
          <li><a href="#">Services</a></li>
          <li><a href="#">Products</a></li>
        </ul>
      </nav>
    </aside>
  </article>
</dialog>
```
```css
nav[role="menubar" i] > ul > li[role=separator] {
  display: block;
}

nav[role="menubar" i] > ul:not(:has(li[role="separator" i])) {
  display: none;
}

@media (min-width: 768px) {
  nav[role="menubar" i] > ul > li[role=separator] {
    display: none;
  }
  nav[role="menubar" i] > ul:not(:has(li[role="separator" i])) {
    display: block;
  }
}
```

Demo: https://jsfiddle.net/e1swg205/

**Comment by dday9** [comment]
An alternative, no-JS approach, would be to leverage the details element. This can leverage the `role="menubar"` attribute, which isn't 100% W3 compliant because technically a details element cannot have this role, but PicoCSS already breaks this rule via the `summary[role="button"]` recommendation:

```html
<details role="menubar" aria-labelledby="menubar-title">
  <summary>
    <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960" width="24px" fill="currentColor" aria-hidden="true" focusable="false">
      <path d="M120-240v-80h720v80H120Zm0-200v-80h720v80H120Zm0-200v-80h720v80H120Z"/>
    </svg>
  </summary>
  <nav aria-label="Website navigation bar">
    <ul aria-label="Brand name">
      <li role="none">
        <strong>Acme Corp</strong>
      </li>
    </ul>
    <ul aria-label="Navigation">
      <li role="none"><a role="menuitem" href="#">About</a></li>
      <li role="none"><a role="menuitem" href="#">Services</a></li>
      <li role="none"><a role="menuitem" href="#">Products</a></li>
    </ul>
  </nav>
</details>
```
```css
details[role="menubar" i] > summary::after {
  display: none;
}

details[role="menubar" i] > summary > nav,
details[role="menubar" i] > nav > ul {
  overflow-x: auto;
}

details[role="menubar" i] > nav,
details[role="menubar" i] > nav > ul {
  flex-direction: column;
}

details[role="menubar" i] > nav > ul > li {
  flex: 0 0 auto;
}

@media (min-width: 576px) {
  details[role="menubar" i] > summary {
    display: none;
  }

  details[role="menubar" i] > nav,
  details[role="menubar" i] > nav > ul {
    flex-direction: row;
  }

  details[role="menubar" i]:not([open])::details-content {
    content-visibility: visible;
  }
}
```

Demo: https://jsfiddle.net/97x5fotz/
